feat(analysis/calculus/mean_value): add generalized "fencing" inequality

urkud opened the pull request:
I'm taking a 3-4 days break from Lean. After this I'll add missing
docstrings to this PR, go over review comments, then formalize
Gronwall inequality based on the general fencing theorem from this PR.

TO CONTRIBUTORS:

Make sure you have:

  * [X] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [ ] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [X] make sure definitions and lemmas are put in the right files
  * [X] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -42,84 +42,177 @@ set_option class.instance_max_depth 120
+* `∥f a∥ ≤ B a`;
+* `B` has a right derivative at every point `[a, b)`;
+* for each `x ∈ [a, b)` the right-side limit superior of the norm of `(f z - f x) / (z - x)`
+  is bounded above by a function `f'` which is strictly less than `B'`.
```
The lemma is true with `limsup` replaced with `liminf`.

urkud replied:
How do I say `liminf f ≤ f'` with filters?

digama0 replied:
You need cluster points. I think `liminf f L <= a` says that for every `b > a` there is a cluster point `c <= b` of the filter `map f L` (where "`F` clusters at `c`" means that for all `A \in F`, and `B \in nhds c`, `A ∩ B != 0`).

urkud replied:
So,  `∀ b > a, ∀ s ∈ L.map f, (s ∩ Iio b).nonempty`?

urkud replied:
Or better `∀ b > a, ∀ s ∈ L, (s ∩ f ⁻¹' (Iio b)).nonempty`?

sgouezel replied:
Something like that, yes. You could also use the `liminf` defined in `src/order/liminf_limsup`, but I don't know how good the API is, and more importantly it only works well in a complete linear order, so you would need to work in `ereal`. While #1703 is not merged, this is not really an option.

urkud replied:
I proved with `∀ b > a, ∀ s ∈ L, (s ∩ f ⁻¹' (Iio b)).nonempty` (not pushed yet, because I made some unrelated modifications; splitting into PRs now).

urkud replied:
Then changed to `filter.frequently`.

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -42,84 +42,177 @@ set_option class.instance_max_depth 120
+  exact inf_le_right
 end
 
+lemma image_norm_le_of_norm_deriv_right_le_deriv_boundary {f' : ℝ → E}
```
You could also give a lemma with the large inequality in `bound`, i.e., replacing it with `(bound : ∀ x ∈ Ico a b, ∥f' x∥ ≤ deriv B x)`. It follows by applying your version to `B x + epsilon (x-a)` and letting `epsilon` tend to `0`.

urkud replied:
Done.

urkud committed:
Adjust to merged branches, use `liminf` instead of `limsup`, add more variations

urkud committed:
Go through dim-1 liminf estimates

urkud committed:
feat(analysis/calculus/mean_value): add generalized "fencing" inequality

This version can be used to deduce, e.g., Gronwall inequality as well
as its generalized version that deals with approximate solutions.

urkud committed:
Fix: use `b ∈ Ioc a c` as a hypothesis for `I??_mem_nhds_within_Iio`

sgouezel commented on src/analysis/calculus/deriv.lean:
```diff
@@ -1210,3 +1223,83 @@ lemma deriv_within_pow (hxs : unique_diff_within_at 𝕜 s x) :
+
+open metric
+
+variables {E : Type u} [normed_group E] [normed_space ℝ E] {f : ℝ → E} {f' : E} {s : set ℝ} {x : ℝ}
```
I think you're overusing the variables mechanism. The statements of the theorems should remain readable, i.e., looking at the theorem only (and not the variables field of the current section) one should be able to understant what the theorem says. Here, it is not the case. I think a good rule of thumb is to avoid `Prop` variables, and state them explicitly in the statements.

urkud replied:
OK, I'll change this.

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -42,84 +42,256 @@ set_option class.instance_max_depth 120
+
+* `f a ≤ B a`;
+* `B` has a right derivative at every point `[a, b)`;
+* for each `x ∈ [a, b)` the right-side limit inferior of the norm of `(f z - f x) / (z - x)`
```
```suggestion
* for each `x ∈ [a, b)` the right-side limit inferior of `(f z - f x) / (z - x)`
```

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -42,84 +42,256 @@ set_option class.instance_max_depth 120
+* `f a ≤ B a`;
+* `B` has a right derivative at every point `[a, b)`;
+* for each `x ∈ [a, b)` the right-side limit inferior of the norm of `(f z - f x) / (z - x)`
+  is bounded above by a function `f'` which is strictly less than `B'`.
```
This does not correspond to the statement of the theorem: you only require `f' < B'` when `f = B`.

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -42,84 +42,256 @@ set_option class.instance_max_depth 120
+    ∃ᶠ z in nhds_within x (Ioi x), (z - x)⁻¹ * (∥f z∥ - ∥f x∥) < r)
+  {B B' : ℝ → ℝ} (ha : ∥f a∥ ≤ B a) (hB : continuous_on B (Icc a b))
+  (hB' : ∀ x ∈ Ico a b, has_deriv_within_at B (B' x) (Ioi x) x)
+  (bound : ∀ x ∈ Ico a b, ∥f x∥ = B x → f' x < B' x) :
```
The docstring does not agree with the statement: you only require `f' < B'` when `||f|| = B`.

urkud replied:
Sometimes I wish we had per-argument docstrings (transforming, e.g., into `<dl>` in the output HTML).

sgouezel reviewed (commented):
The mathematical content is great, but documentation could still be improved. Notably, the file docstring of `mean_value.lean` should be updated. Also, there are various similar statements, close but slightly different. It would help to comment a little bit on the differences and the different use cases.

urkud committed:
Update src/analysis/calculus/mean_value.lean

Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>

urkud committed:
Drop `Prop`-valued `variables`, add some docs

urkud committed:
Merge branch 'fence' of git://github.com/leanprover-community/mathlib into fence

urkud committed:
More docstrings

urkud committed:
Merge branch 'master' into fence

sgouezel commented on src/analysis/calculus/deriv.lean:
```diff
@@ -1210,3 +1223,125 @@ lemma deriv_within_pow (hxs : unique_diff_within_at 𝕜 s x) :
+
+section real
+
+variables {f : ℝ → ℝ} {f' : ℝ} {s : set ℝ} {x : ℝ} (hs : x ∉ s) (hf : has_deriv_within_at f f' s x)
```
Could you put the three `Prop` assumptions in the statements to help the readers?

urkud replied:
Done.

sgouezel commented on src/analysis/calculus/deriv.lean:
```diff
@@ -1210,3 +1223,125 @@ lemma deriv_within_pow (hxs : unique_diff_within_at 𝕜 s x) :
+see `has_deriv_within_at.limsup_norm_slope_le'`. -/
+lemma has_deriv_within_at.limsup_norm_slope_le
+  (hf : has_deriv_within_at f f' s x) (hr : ∥f'∥ < r) :
+  ∀ᶠ z in nhds_within x (s \ {x}), ∥z - x∥⁻¹ * ∥f z - f x∥ < r :=
```
I don't get why you need to exclude `x`: at `x`, the left-hand side is `0`, and the right hand side is positive thanks to `hr`, so the conclusion also holds.

urkud replied:
Good catch: I just copied from the general (non-`norm`) version. Fixed.

urkud committed:
Merge branch 'master' into fence

urkud committed:
Merge branch 'master' into fence

urkud committed:
Drop `Prop`-valued `variables`, drop assumption `x ∉ s`.

mergify[bot] committed:
Merge branch 'master' into fence
